### PR TITLE
Changes to support Numpy >= 1.24

### PR DESCRIPTION
**Pull request by shwina**
## Description
Closes https://github.com/rapidsai/cudf/issues/13301

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/__init__.py
Comment (shwina) [inline python/cudf/cudf/__init__.py:100]: This hack will go away

**Comment by vyasr** [comment]
Can you remove the numpy upper bound pinnings in dependencies.yaml and the cudf meta.yaml? Then in the CUDA 12.0 CI let's make sure that at least one run installs numpy 1.24 and still passes tests.

File: python/cudf/cudf/__init__.py
Comment (vyasr) [reply python/cudf/cudf/__init__.py:100]: I think it can be removed from this PR itself. Now that we've verified numpy 1.24 support, I would recommend removing the numba-related changes in this PR that you're using in order to allow running numba 0.57 (which is necessary to use numpy 1.24). We'll still get it tested because our CUDA 12 wheel builds will patch us to use 0.57 anyway (but with CUDA 12 we don't use cubinlinker/ptxcompiler so we don't need any edits for those). Then when we bump our numba to 0.57 tests should pass thanks to this PR.

File: python/cudf/cudf/tests/test_csv.py
Comment (vyasr) [reply python/cudf/cudf/tests/test_csv.py:153]: Could you elaborate? Is this a task you want to accomplish in this PR?

File: python/cudf/cudf/tests/test_column.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_column.py:401]: I guess it doesn't matter what value we choose here? Just wondering if it's important to use 453-256.

File: python/cudf/cudf/tests/test_numerical.py
Comment (vyasr) [inline python/cudf/cudf/tests/test_numerical.py:197]: Instead of applying this to the whole test, can we just wrap the `pd.to_numeric` call? This doesn't affect the `cudf.to_numeric` call, does it?

Also, should we be handling the warning conditionally? i.e. I assuming this happens when trying to downcast a signed to an unsigned type or something?

File: python/cudf/cudf/__init__.py
Comment (vyasr) [reply python/cudf/cudf/__init__.py:100]: I have one other question on this PR, would wait to make changes here until everything else is resolved in case you need to run more tests locally.

**Review by vyasr — COMMENTED** [review]
Success! https://github.com/rapidsai/cudf/actions/runs/4953167580/jobs/8860521944?pr=13325#step:10:828 shows that numpy 1.24 is now getting installed into the 12.0.1 wheel build, and tests are passing

File: python/cudf/cudf/tests/test_column.py
Comment (shwina) [reply python/cudf/cudf/tests/test_column.py:401]: I don't think it matters. I use `255` just because it's `np.iinfo(uint8).max'

File: python/cudf/cudf/utils/queryutils.py
Comment (wence-) [inline python/cudf/cudf/utils/queryutils.py:140]: This is incorrect, `hash` returns in the semi-open interval `[-2**63, 2**63)`, but `abs` folds this to the closed interval `[0, 2**63]` (so you alias just shy of 50% of the values). Instead, you want to shift, I suspect, and then you don't need numpy in the loop at all:

```suggestion
    funcid = f"queryexpr_{hash(expr) + 2**63:x}"
```

That said, strings are hashable, so this seems like a weird way of constructing a cache key (it's somehow deliberately making it more likely that you get hash collisions and produce the wrong value).

I would have thought that this would do the trick:

```
@functools.cache
def query_compile(expr):
    name  = "queryexpr" # these are only looked up locally so names can collide
    info = query_parser(expr)
    fn = query_builder(info, name)
    args = info["args"]
    devicefn = cudf.jit(device=True)(fn)
    kernel = _wrap_query_expr(f"kernel_{name}", devicefn, args)
    info["kernel"] = kernel
    return info
```

**Review by wence- — CHANGES_REQUESTED** [review]
It seems like there are some strange things going on in the changes in the tests?

File: python/cudf/cudf/tests/test_csv.py
Comment (shwina) [reply python/cudf/cudf/tests/test_csv.py:153]: `np.uint8(-1)` in particular raises a deprecation warning. But the reason I gave up on this test was because I couldn't figure out why this was happening:

```python
In [2]: np.array([-1]).astype("uint64")
Out[2]: array([18446744073709551615], dtype=uint64)

In [3]: np.array([18446744073709551615]).astype("uint64")
Out[3]: array([18446744073709551615], dtype=uint64)

In [4]: np.array([-1, 18446744073709551615]).astype("uint64")
<ipython-input-4-03014ed268fc>:1: RuntimeWarning: invalid value encountered in cast
  np.array([-1, 18446744073709551615]).astype("uint64")
Out[4]: array([18446744073709551615,                    0], dtype=uint64)
```

I've gone ahead and filtered out that warning from this test.

File: python/cudf/cudf/tests/test_unaops.py
Comment (shwina) [reply python/cudf/cudf/tests/test_unaops.py:87]: Well:

```
In [2]: -np.uint16(1)
<ipython-input-2-5156426c8f88>:1: RuntimeWarning: overflow encountered in scalar negative
  -np.uint16(1)
Out[2]: 65535
```

Should we just go ahead and ignore that warning in this test? (I've resorted to doing that in most other cases)

**Comment by vyasr** [comment]
@shwina Is this PR waiting on a re-review from me and/or @wence- ?

**Comment by shwina** [comment]
@vyasr just resolved the conflicts here. Yeah I think it could use a quick skim

**Pull request merged**
